hikari/impl/rest.py

```diff
@@ -1446,26 +1446,25 @@ async def delete_messages(
         channel: snowflakes.SnowflakeishOr[channels_.TextableChannel],
         messages: typing.Union[
             snowflakes.SnowflakeishOr[messages_.PartialMessage],
-            snowflakes.SnowflakeishIterable[messages_.PartialMessage],
+            typing.Iterable[snowflakes.SnowflakeishOr[messages_.PartialMessage]],
+            typing.AsyncIterable[snowflakes.SnowflakeishOr[messages_.PartialMessage]],
         ],
         /,
         *other_messages: snowflakes.SnowflakeishOr[messages_.PartialMessage],
     ) -> None:
         route = routes.POST_DELETE_CHANNEL_MESSAGES_BULK.compile(channel=channel)
 
-        pending: typing.List[snowflakes.SnowflakeishOr[messages_.PartialMessage]] = []
         deleted: typing.List[snowflakes.SnowflakeishOr[messages_.PartialMessage]] = []
 
-        if isinstance(messages, typing.Iterable):  # Syntactic sugar. Allows to use iterables
-            pending.extend(messages)
-
+        if isinstance(messages, typing.AsyncIterable):
+            iterator = iterators.NOOPLazyIterator(messages)
```

Comment:
maybe add a `LazyIterator.__new__` that can return the appropriate iterator?

iterable -> flat lazy iterator
async iterable -> noop iterator
lazy iterator -> identical

Reply:
Probably outside the scope of this PR